cpp/src/datetime/datetime_ops.cu

```diff
@@ -76,9 +76,14 @@ struct extract_component_operator {
 
     if (time_since_midnight.count() < 0) { time_since_midnight += days(1); }
 
-    auto hrs_  = duration_cast<hours>(time_since_midnight);
-    auto mins_ = duration_cast<minutes>(time_since_midnight - hrs_);
-    auto secs_ = duration_cast<seconds>(time_since_midnight - hrs_ - mins_);
+    auto hrs_       = duration_cast<hours>(time_since_midnight);
+    auto mins_      = duration_cast<minutes>(time_since_midnight - hrs_);
+    auto secs_      = duration_cast<seconds>(time_since_midnight - hrs_ - mins_);
+    auto millisecs_ = duration_cast<milliseconds>(time_since_midnight - hrs_ - mins_ - secs_);
+    auto microsecs_ =
+      duration_cast<microseconds>(time_since_midnight - hrs_ - mins_ - secs_ - millisecs_);
+    auto nanosecs_ = duration_cast<nanoseconds>(time_since_midnight - hrs_ - mins_ - secs_ -
+                                                millisecs_ - microsecs_);
```

Comment:
This function is inefficient. All these values are computed but only one is used. So each of them should be computed inside the `switch` statement.

Reply:
So I think this should be changed: Only compute the value if needed.

Reply:
I see your point, @ttnghia, but there are transitive dependencies. (`hrs_` is required for all the following values. `mins_` is required for everything that follows it, etc.)

How do you propose to reduce code amplification?

Reply:
I tried addressing the inefficient calculations this way: https://github.com/rapidsai/cudf/pull/11911/commits/16e7b8f12c3d31c37501cc64b73a95491ba1da3c let me know what you both think of it.

Reply:
Oh no, this dependent chain is even worse. Sorry @galipremsagar please reverse it :(.

Reply:
Done 👍